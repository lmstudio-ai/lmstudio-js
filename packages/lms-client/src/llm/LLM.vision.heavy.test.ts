import { existsSync } from "fs";
import { unlink, writeFile } from "fs/promises";
import { tmpdir } from "os";
import { join } from "path";
import { type LLM, LMStudioClient } from "../index.js";
import { ensureHeavyTestsEnvironment, llmTestingQwenVL } from "../shared.heavy.test.js";

describe("LLM with vision", () => {
  let client: LMStudioClient;
  let model: LLM;
  let filePath: string;
  let imageContent: ArrayBuffer;
  beforeAll(async () => {
    client = new LMStudioClient();
    await ensureHeavyTestsEnvironment(client);
    imageContent = await fetch(
      "https://www.w3.org/People/mimasa/test/imgformat/img/w3c_home.jpg",
    ).then(response => response.arrayBuffer());
    filePath = join(tmpdir(), "w3c_home_" + Date.now() + ".jpg");
    await writeFile(filePath, new Uint8Array(imageContent));
  }, 60_000);
  afterAll(async () => {
    if (existsSync(filePath)) {
      await unlink(filePath);
    }
  });
  beforeEach(async () => {
    model = await client.llm.model(llmTestingQwenVL, {
      verbose: false,
      config: {
        llamaKCacheQuantizationType: "f32",
        llamaVCacheQuantizationType: "f32",
      },
    });
  }, 60_000);
  it("should work with base64 upload", async () => {
    const image = await client.files.prepareImageBase64(
      "w3c_home.jpg",
      Buffer.from(new Uint8Array(imageContent)).toString("base64"),
    );
    const result = await model.respond(
      { content: "What do you see?", images: [image] },
      { temperature: 0, maxTokens: 10 },
    );
    expect(result.content).toMatchSnapshot();
    expect(result.stats).toMatchSnapshot({
      numGpuLayers: expect.any(Number),
      timeToFirstTokenSec: expect.any(Number),
      tokensPerSecond: expect.any(Number),
    });
    expect(result.modelInfo).toMatchSnapshot({
      identifier: expect.any(String),
      instanceReference: expect.any(String),
      modelKey: expect.any(String),
    });
    expect(typeof (await image.getFilePath())).toBe("string");
    expect(typeof image.identifier).toBe("string");
    expect(image.sizeBytes).toMatchInlineSnapshot(`2165`);
    expect(image.type).toBe("image");
  });
  it("should work with file upload", async () => {
    const image = await client.files.prepareImage(filePath);
    const result = await model.respond(
      { content: "What do you see?", images: [image] },
      { temperature: 0, maxTokens: 10 },
    );
    expect(result.content).toMatchSnapshot();
    expect(result.stats).toMatchSnapshot({
      numGpuLayers: expect.any(Number),
      timeToFirstTokenSec: expect.any(Number),
      tokensPerSecond: expect.any(Number),
    });
    expect(result.modelInfo).toMatchSnapshot({
      identifier: expect.any(String),
      instanceReference: expect.any(String),
      modelKey: expect.any(String),
    });
    expect(typeof (await image.getFilePath())).toBe("string");
    expect(typeof image.identifier).toBe("string");
    expect(image.sizeBytes).toMatchInlineSnapshot(`2165`);
    expect(image.type).toBe("image");
  });
});
