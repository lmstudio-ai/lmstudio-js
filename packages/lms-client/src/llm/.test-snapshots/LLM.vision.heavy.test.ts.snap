// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LLM with vision should work with base64 upload 1`] = `"I see the W3C logo, which"`;

exports[`LLM with vision should work with base64 upload 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 10,
  "promptTokensCount": 27,
  "stopReason": "maxPredictedTokensReached",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 37,
}
`;

exports[`LLM with vision should work with base64 upload 3`] = `
{
  "architecture": "qwen2vl",
  "contextLength": 4096,
  "displayName": "Qwen2 VL 2B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "2B",
  "path": "lmstudio-community/Qwen2-VL-2B-Instruct-GGUF/Qwen2-VL-2B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 3647162336,
  "trainedForToolUse": false,
  "type": "llm",
  "vision": true,
}
`;

exports[`LLM with vision should work with file upload 1`] = `"I see the W3C logo, which"`;

exports[`LLM with vision should work with file upload 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 10,
  "promptTokensCount": 27,
  "stopReason": "maxPredictedTokensReached",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 37,
}
`;

exports[`LLM with vision should work with file upload 3`] = `
{
  "architecture": "qwen2vl",
  "contextLength": 4096,
  "displayName": "Qwen2 VL 2B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "2B",
  "path": "lmstudio-community/Qwen2-VL-2B-Instruct-GGUF/Qwen2-VL-2B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 3647162336,
  "trainedForToolUse": false,
  "type": "llm",
  "vision": true,
}
`;
